From bacb38d863d8ccca61b23fb16d3c252c02691723 Mon Sep 17 00:00:00 2001
From: Sultan Alsawaf <sultan@kerneltoast.com>
Date: Tue, 18 Jul 2023 21:26:04 -0700
Subject: [PATCH] mm: Micro-optimize PID maps output for arm64

Android and various applications in Android need to read PID map data in
order to work. Some processes can contain over 10,000 mappings, which
results in lots of time wasted on simply generating strings. This wasted
time adds up, especially in the case of Unity-based games, which utilize
the Boehm garbage collector. A game's main process typically has well over
10,000 mappings due to the loaded textures, and the Boehm GC reads PID maps
several times a second. This results in over 100,000 map entries being
printed out per second, so micro-optimization here is important.

The primary micro-optimizations here assume that there are no more than 40
bits in the virtual address space, hence the CONFIG_ARM64_VA_BITS check.
Arm64 uses a virtual address size of 39 bits, so this perfectly covers it.

This also removes padding used to beautify PID map output to further speed
up reads and reduce the amount of bytes printed, and optimizes the dentry
path retrieval for file-backed mappings. Note, however, that the trailing
space at the end of the line for non-file-backed mappings cannot be
omitted, as it breaks some PID map parsers.

This still retains insignificant leading zeros from printed hex values to
maintain the current output format in order to avoid breaking parsers.

This patch version is for 5.10 kernels and combines a subsequent patch,
"mm: Eliminate d_path_outlen() and further speed up show_map_vma()". It
also doesn't contain the unsafe memcpy() optimization to prepend_name().

Signed-off-by: Sultan Alsawaf <sultan@kerneltoast.com>
---
 fs/proc/task_mmu.c | 194 +++++++++++++++++++++++++++++++++++++--------
 1 file changed, 159 insertions(+), 35 deletions(-)

Index: common/fs/proc/task_mmu.c
===================================================================
--- common/fs/proc/task_mmu.c
+++ common/fs/proc/task_mmu.c
@@ -240,24 +240,117 @@ static int do_maps_open(struct inode *in
 				sizeof(struct proc_maps_private));
 }
 
-static void show_vma_header_prefix(struct seq_file *m,
-				   unsigned long start, unsigned long end,
-				   vm_flags_t flags, unsigned long long pgoff,
-				   dev_t dev, unsigned long ino)
+#define print_vma_hex10(out, val, clz_fn) \
+({									\
+	const typeof(val) __val = val;					\
+	char *const __out = out;					\
+	size_t __len;							\
+									\
+	if (__val) {							\
+		__len = (sizeof(__val) * 8 - clz_fn(__val) + 3) / 4;	\
+		switch (__len) {					\
+		case 10:						\
+			__out[9] = hex_asc[(__val >>  0) & 0xf];	\
+			__out[8] = hex_asc[(__val >>  4) & 0xf];	\
+			__out[7] = hex_asc[(__val >>  8) & 0xf];	\
+			__out[6] = hex_asc[(__val >> 12) & 0xf];	\
+			__out[5] = hex_asc[(__val >> 16) & 0xf];	\
+			__out[4] = hex_asc[(__val >> 20) & 0xf];	\
+			__out[3] = hex_asc[(__val >> 24) & 0xf];	\
+			__out[2] = hex_asc[(__val >> 28) & 0xf];	\
+			__out[1] = hex_asc[(__val >> 32) & 0xf];	\
+			__out[0] = hex_asc[(__val >> 36) & 0xf];	\
+			break;						\
+		case 9:							\
+			__out[8] = hex_asc[(__val >>  0) & 0xf];	\
+			__out[7] = hex_asc[(__val >>  4) & 0xf];	\
+			__out[6] = hex_asc[(__val >>  8) & 0xf];	\
+			__out[5] = hex_asc[(__val >> 12) & 0xf];	\
+			__out[4] = hex_asc[(__val >> 16) & 0xf];	\
+			__out[3] = hex_asc[(__val >> 20) & 0xf];	\
+			__out[2] = hex_asc[(__val >> 24) & 0xf];	\
+			__out[1] = hex_asc[(__val >> 28) & 0xf];	\
+			__out[0] = hex_asc[(__val >> 32) & 0xf];	\
+			break;						\
+		default:						\
+			__out[7] = hex_asc[(__val >>  0) & 0xf];	\
+			__out[6] = hex_asc[(__val >>  4) & 0xf];	\
+			__out[5] = hex_asc[(__val >>  8) & 0xf];	\
+			__out[4] = hex_asc[(__val >> 12) & 0xf];	\
+			__out[3] = hex_asc[(__val >> 16) & 0xf];	\
+			__out[2] = hex_asc[(__val >> 20) & 0xf];	\
+			__out[1] = hex_asc[(__val >> 24) & 0xf];	\
+			__out[0] = hex_asc[(__val >> 28) & 0xf];	\
+			__len = 8;					\
+		break;						\
+		}							\
+	} else {							\
+		*(u64 *)__out = U64_C(0x3030303030303030);		\
+		__len = 8;						\
+	}								\
+											\
+	__len;								\
+})
+
+#define print_vma_hex2(out, val) \
+({									\
+	const typeof(val) __val = val;					\
+	char *const __out = out;					\
+									\
+	__out[1] = hex_asc[(__val >>  0) & 0xf];			\
+	__out[0] = hex_asc[(__val >>  4) & 0xf];			\
+									\
+	2;								\
+})
+
+static int show_vma_header_prefix(struct seq_file *m, unsigned long start,
+				unsigned long end, vm_flags_t flags,
+				unsigned long long pgoff, dev_t dev,
+				unsigned long ino)
 {
-	seq_setwidth(m, 25 + sizeof(void *) * 6 - 1);
-	seq_put_hex_ll(m, NULL, start, 8);
-	seq_put_hex_ll(m, "-", end, 8);
-	seq_putc(m, ' ');
-	seq_putc(m, flags & VM_READ ? 'r' : '-');
-	seq_putc(m, flags & VM_WRITE ? 'w' : '-');
-	seq_putc(m, flags & VM_EXEC ? 'x' : '-');
-	seq_putc(m, flags & VM_MAYSHARE ? 's' : 'p');
-	seq_put_hex_ll(m, " ", pgoff, 8);
-	seq_put_hex_ll(m, " ", MAJOR(dev), 2);
-	seq_put_hex_ll(m, ":", MINOR(dev), 2);
-	seq_put_decimal_ull(m, " ", ino);
-	seq_putc(m, ' ');
+	size_t len;
+	char *out;
+
+	/* Set the overflow status to get more memory if there's no space */
+	if (seq_get_buf(m, &out) < 65) {
+		seq_commit(m, -1);
+		return -ENOMEM;
+	}
+
+	/* Supports printing up to 40 bits per virtual address */
+	BUILD_BUG_ON(CONFIG_ARM64_VA_BITS > 40);
+
+	len = print_vma_hex10(out, start, __builtin_clzl);
+
+	out[len++] = '-';
+
+	len += print_vma_hex10(out + len, end, __builtin_clzl);
+
+	out[len++] = ' ';
+	out[len++] = "-r"[!!(flags & VM_READ)];
+	out[len++] = "-w"[!!(flags & VM_WRITE)];
+	out[len++] = "-x"[!!(flags & VM_EXEC)];
+	out[len++] = "ps"[!!(flags & VM_MAYSHARE)];
+	out[len++] = ' ';
+
+	len += print_vma_hex10(out + len, pgoff, __builtin_clzll);
+
+	out[len++] = ' ';
+
+	len += print_vma_hex2(out + len, MAJOR(dev));
+
+	out[len++] = ':';
+
+	len += print_vma_hex2(out + len, MINOR(dev));
+
+	out[len++] = ' ';
+
+	len += num_to_str(&out[len], 20, ino, 0);
+
+	out[len++] = ' ';
+
+	m->count += len;
+	return 0;
 }
 
 static void
@@ -285,7 +378,8 @@ show_map_vma(struct seq_file *m, struct
 
 	__fold_filemap_fixup_entry(&((struct proc_maps_private *)m->private)->iter, &end);
 
-	show_vma_header_prefix(m, start, end, flags, pgoff, dev, ino);
+	if (show_vma_header_prefix(m, start, end, flags, pgoff, dev, ino))
+		return;
 	if (mm)
 		anon_name = anon_vma_name(vma);
 
@@ -294,16 +388,46 @@ show_map_vma(struct seq_file *m, struct
 	 * special [heap] marker for the heap:
 	 */
 	if (file) {
-		seq_pad(m, ' ');
+
 		/*
 		 * If user named this anon shared memory via
 		 * prctl(PR_SET_VMA ..., use the provided name.
 		 */
-		if (anon_name)
+		if (anon_name) {
+			seq_pad(m, ' ');
 			seq_printf(m, "[anon_shmem:%s]", anon_name->name);
-		else
-			seq_file_path(m, file, "\n");
-		goto done;
+			goto done;
+		}
+		else {
+			char *buf;
+			size_t size = seq_get_buf(m, &buf);
+
+			/*
+			* This won't escape newline characters from the path. If a
+			* program uses newlines in its paths then it can kick rocks.
+			*/
+			if (size > 1) {
+				char *p;
+
+				p = d_path(&file->f_path, buf, size);
+				if (!IS_ERR(p)) {
+					size_t len;
+
+					/* Minus one to exclude the NUL character */
+					len = size - (p - buf) - 1;
+					if (likely(p > buf))
+						memmove(buf, p, len);
+						buf[len] = '\n';
+						seq_commit(m, len + 1);
+						return;
+				}
+			}
+
+			/* Set the overflow status to get more memory */
+			seq_commit(m, -1);
+			return;
+		}
+
 	}
 
 	if (vma->vm_ops && vma->vm_ops->name) {
@@ -315,31 +439,31 @@ show_map_vma(struct seq_file *m, struct
 	name = arch_vma_name(vma);
 	if (!name) {
 		if (!mm) {
-			name = "[vdso]";
-			goto done;
+			seq_write(m, "[vdso]\n", 7);
+			return;
 		}
 
 		if (vma_is_initial_heap(vma)) {
-			name = "[heap]";
-			goto done;
+			seq_write(m, "[heap]\n", 7);
+			return;
 		}
 
 		if (vma_is_initial_stack(vma)) {
-			name = "[stack]";
-			goto done;
+			seq_write(m, "[stack]\n", 8);
+			return;
 		}
 
 		if (anon_name) {
-			seq_pad(m, ' ');
-			seq_printf(m, "[anon:%s]", anon_name->name);
+			seq_write(m, "[anon:", 6);
+			seq_puts(m, anon_name->name);
+			seq_write(m, "]\n", 2);
+			return;
 		}
 	}
 
 done:
-	if (name) {
-		seq_pad(m, ' ');
+	if (name)
 		seq_puts(m, name);
-	}
 	seq_putc(m, '\n');
 }
 
